Entities related to Entities of the same type

benbosman opened the pull request:
## Description
This resolves a bug if the left type and the right type of a relationship is identical. That setup was causing the virtual metadata to be populated incorrectly and the counts to be defined incorrectly.

## Instructions for Reviewers
The only 2 changes for the virtual metadata are in dspace-api/src/main/java/org/dspace/content/RelationshipMetadataServiceImpl.java to verify not only on the type, but also on the item's UUID
The counts are solved by defining whether the left or right side should be verified

There are already many existing ITs which still succeed.
A new type of relationship has been defined in the test now where the left type and the right type of a relationship is identical

## Checklist
- [X] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [X] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide)
- [X] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [X] My PR passes all tests and includes new/updated Unit or Integration Tests for any bug fixes, improvements or new features. A few reminders about what constitutes good tests:
    * Include tests for different user types, including: (1) Anonymous user, (2) Logged in user (non-admin), and (3) Administrator.
    * Include tests for known error scenarios and error codes (e.g. `400 Bad Request`, `401 Unauthorized`, `403 Forbidden`, `404 Not Found`, etc)
    * For bug fixes, include a test that reproduces the bug and proves it is fixed. For clarity, it may be useful to provide the test in a separate commit from the bug fix.
- [X] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/master/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [X] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

benbosman committed:
Support for a relationship from type A to type A (parent/child of same type)

Raf-atmire committed:
[Task 70231] Added orgUnit x orgUnit relationshiptype and test to prove the labels in a circular relationship

Raf-atmire committed:
[Task 70401] adding failing IT to prove that count is wrong with parent and child OrgUnits in the findByLabel

Raf-atmire committed:
[Task 70401] fixed the RelationshipService#countByItemAndRelationshipType to keep the isLeft parameter into account

benbosman committed:
Merge pull request #19 from atmire/w2p-70401_count-queries-circular-entities

W2p 70401 count queries circular entities

Raf-atmire committed:
[Task 70419] fixed a bug where the RelationshipService creation did not check the max cardinality properly

MarieVerdonck committed:
Git merge https://github.com/atmire/DSpace.git / master
* Conflict file: RelationshipRestRepositoryIT

tdonohue commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/RelationshipRestRepositoryIT.java:
```diff
@@ -105,60 +116,60 @@ public void setUp() throws Exception {
-                                          .withName("Parent Community")
-                                          .build();
+                .withName("Parent Community")
+                .build();
```
@benbosman : Just a quick note that this PR needs a rebase & cleanup.  It looks like nearly all the changes to `RelationshipRestRepositoryIT` are **space changes** (where the IDE seems to have made a lot of unrelated alignment changes).  If you could revert those space changes it'd make this PR easier to review (and it might even fall into the "1 approval" category).

benbosman commented:
@tdonohue the branch has been rebased

paulo-graca commented on dspace-api/src/main/java/org/dspace/content/dao/impl/RelationshipDAOImpl.java:
```diff
@@ -201,18 +201,24 @@ public int countRows(Context context) throws SQLException {
-    public int countByItemAndRelationshipType(Context context, Item item, RelationshipType relationshipType)
-            throws SQLException {
+    public int countByItemAndRelationshipType(Context context, Item item, RelationshipType relationshipType,
+                                              boolean isLeft) throws SQLException {
```
I don't specially like the approach of passing parameters that in my perspective describe the property of an object. In this particular case the object Relationship. The concept of Left and Right item are always subjective and relative to the person that defines it. I think a different approach that could be considered in the future is to probably considers an Item not for the location/position which is it, but if it's itself in a relation verification.

But I understand the reason for having this boolean here. And considering the required effort for a different approach, I'm ok with having it this way in this version.

tdonohue replied:
I agree with @paulo-graca 's comment here.  It seems like this `isLeft` is needed, but it's an awkward implementation...the left/right concept is turning out to be quite confusing (and it is subjective).  I'm not sure what to replace it with though, as that would take more detailed discussion that would likely need to be delayed for post 7.0

benbosman replied:
The Left/Right contents is indeed not pre-defined, and just dependent on how the relation is configured. The concept of Left/Right was chosen to make it easier if you draw the relationship on paper.
It's not an easy concept to support any relationship in code, but feel free to create a ticket with suggestions to make this easier to understand.
For 7.0 it's indeed not realistic to adjust this

paulo-graca reviewed (approved):
My comment goes in the way of a repeating speech that we need to take a second/third look at the solution we found for relationships. From the first moment it was presented it raised doubts and some discussion. We are entering in a relative or subjective realm to the person that defines the relationship. 
The presented solution addresses the issue, but by adopting it we are adding complexity to the system (this is not particular to this PR, it's related with the current Relationship model). I suggest to consider a new approach perhaps it a further release or version of DSpace.

tdonohue commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/RelationshipRestRepositoryIT.java:
```diff
@@ -2482,5 +2500,160 @@ public void putRelationshipWithJson() throws Exception {
+        context.restoreAuthSystemState();
+        // Here we create our first Relationship to the Publication to give it a dc.contributor.author virtual
+        // metadata field.
+        MvcResult mvcResult = getClient(adminToken).perform(post("/api/core/relationships")
```
This `post()` and all the others in this IT do not seem to be cleaning up after themselves.  Remember we should be using a try/finally, using an `.andDo()` to capture the ID of the created object (I suspect this code was just written prior to that best practice being established).  In any case, here's an example: https://github.com/DSpace/DSpace/blob/main/dspace-server-webapp/src/test/java/org/dspace/app/rest/ScriptRestRepositoryIT.java#L195-L207

benbosman replied:
This has been updated

tdonohue commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/RelationshipRestRepositoryIT.java:
```diff
@@ -2482,5 +2500,160 @@ public void putRelationshipWithJson() throws Exception {
+        context.restoreAuthSystemState();
+        // Here we create our first Relationship to the Publication to give it a dc.contributor.author virtual
+        // metadata field.
+        getClient(adminToken).perform(post("/api/core/relationships")
```
Again, `post` should cleanup after itself

benbosman replied:
This has been updated

tdonohue commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/RelationshipRestRepositoryIT.java:
```diff
@@ -2482,5 +2500,160 @@ public void putRelationshipWithJson() throws Exception {
+        context.restoreAuthSystemState();
+        // Here we create our first Relationship to the Publication to give it a dc.contributor.author virtual
+        // metadata field.
+        getClient(adminToken).perform(post("/api/core/relationships")
```
Again, `post` should cleanup after itself

benbosman replied:
This has been updated

tdonohue commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/RelationshipRestRepositoryIT.java:
```diff
@@ -2482,5 +2500,160 @@ public void putRelationshipWithJson() throws Exception {
+                                                  .getID()))
+                             .andExpect(status().isCreated());
+
+        getClient(adminToken).perform(post("/api/core/relationships")
```
Again, `post` should cleanup after itself

benbosman replied:
This has been updated

tdonohue reviewed (changes requested):
@benbosman : I gave this a code review today.  The main concepts are OK, but the tests need some minor cleanup (to ensure `post` calls are cleaning up content they create).  See more comments inline below.

MarieVerdonck committed:
Merge branch 'master' into circular-entities

MarieVerdonck committed:
74947: Missing cleanup after post relationship creation

MarieVerdonck committed:
74947: Adjusting HandleDAOImplTest suffixes since this fails in CI tests:
because of an existing Collection with the same handle already existing in test DB
(problem does not occur running all tests locally)

benbosman commented:
@tdonohue Marie has completed the updates

She did notice HandleDAOImplTest stopped working, even though this PR doesn't touch any API tests. She confirmed that the test itself works. This most likely implies there's another test in the API which doesn't clean up data correctly. Due to the randomness in the tests, it suddenly overlapped. This has been temporarily bypassed in https://github.com/DSpace/DSpace/pull/2746/commits/d70e121e93a31dba899850aa399eca35aa7dbf82 because it's not easy to identify where the problem resides and it's out of scope for this PR.

I assume it's best to create a ticket for searching the TEST which creates the collection and doesn't remove it?

tdonohue reviewed (approved):
👍 Looks good to me now.  Thanks @benbosman and @MarieVerdonck !
